tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -288,27 +313,66 @@ class UnsortedSegmentSumOp : public OpKernel {
     auto output_flat = output->flat_outer_dims<T>();
 
     auto data_ptr = data.template flat<T>().data();
-    functor::UnsortedSegmentSumFunctor<Device, T, Index>()(
-        context, context->template eigen_device<Device>(), output_rows,
+    ReductionFunction(context, context->template eigen_device<Device>(), output_rows,
         segment_ids.shape(), segment_flat, data.NumElements(), data_ptr,
         output_flat);
   }
+private:
+  Functype ReductionFunction;
 };
 
-#define REGISTER_CPU_UNSORTED_KERNELS(type, index_type)                \
+template <typename Device, class T, class Index>
+class UnsortedSegmentSumOp
+  : public UnsortedSegmentBaseOp<Device, T, Index,
+                                   functor::UnsortedSegmentSumFunctor<Device, T, Index>> {
+public:
+  explicit UnsortedSegmentSumOp(OpKernelConstruction* context)
+    : UnsortedSegmentBaseOp<Device, T, Index, functor::UnsortedSegmentSumFunctor<Device, T, Index>>(context, functor::UnsortedSegmentSumFunctor<Device, T, Index>()) {}
+};
+
+template <typename Device, class T, class Index>
+class UnsortedSegmentMaxOp
+  : public UnsortedSegmentBaseOp<Device, T, Index,
+                functor::UnsortedSegmentMaxFunctor<Device, T, Index>> {
```

Comment:
@andydavis1 , unfortunately I don't see a way to implement this easily.
I've tried to introduce a BaseFunctor, to pass the specialized Functors as argument for the BaseOp, unfortunately it will only use the BaseFunctors operator, instead of the specialized max and sum operators.
You can introduce pointers to use the specialized functors as here:
https://github.com/nikste/tensorflow/blob/TF-test2/tensorflow/core/kernels/segment_reduction_ops.cc#L316
This compiles, but breaks upon execution in python.
Maybe you have a suggestion how this could be done better?